ls1intum/Artemis
# `Development`: Refactor file service

### Checklist
#### General


#### Server
- [x] I added multiple tests (Spring) related to the features (with a high test coverage).

### Motivation and Context


The `FileService` grew over many years. It not only works mostly with Strings as file paths or URI paths, but also contains lots of duplicated code or bad practices. It also contains code that does not fit the service's context.

closes #6541, #5793

### Description

I moved the method handling model updates to a dedicated service and extracted separate methods from it for the respective actions: Deletion, Update, and Creation.
I also moved methods responsible for the conversion between URI paths (publicPaths) and file paths to a dedicated service.
The service now only contains methods relating to persisting files. We only accept `URI` and `Path` types as input or output in regard to file paths.
Due to that, I had to create a serializer for Hazlecast as I did not want to make an exception for the caching method.
I further renamed multiple methods to make it more clear what they do, suppressed and replaced the Usage of `Files` when writing files as it doesn't make sure the directories exist and simplified much of the structure in the service.

As a result, many file changes only relate to `moved to a different service`, `method renamed` or `String to path/URI`.
For the FileService, I recommend not looking at the file change but rather comparing the old service with the new one.

### Steps for Testing

Create, Update (with a new image) and Delete. Whenever possible open the file in a new tab and check that it gets created/updated/deleted.
- Lecture Attachments (Upload a File to a lecture, change it and then delete the attachment and/or the lecture)
- File Attachment Units & Lecture slides (Create a file attachment unit (PDF), update it with a new file (PDF) and then delete it)
- Course Icons (Create a course with an icon, change that icon and then delete the course)
- File Upload Submissions (submit a file upload submission, update that submission with a new file and then delete the submission by deleting the exercise and/or the course)
- Exam User Signatures (Exam iOS App)
- Exam User Images (Exam iOS App)
- Drag Item & Drag and Drop Background (Create a DnD quiz with background and image drag items, update the images of background and drag items, delete the quiz)

We also need to test the **Export**. However, they all use the same code at its core. Just test any export (e.g. the course exam export of any existing exam). If you want, you can also test additional exports.

Next, we need to test the **Plagiarism detection report creation**.

- Create an exercise (e.g. Programming Exercise) that supports Plagiarism Detection, create multiple submissions
- Click `Check Plagiarism` in the exercise's detail view
- Hit that caret/arrow on the left to extend the menu, check "Generate JPlag Report"
- Detect plagiarism, wait for it to complete and the file to download

### Review Progress 






#### Performance Review
#### Code Review
Many file changes only relate to `moved to a different service`, `method renamed` or `String to path/URI`.
For the FileService, I recommend not looking at the file change but rather comparing the old service with the new one.
Because I touched lots of path conversions, I triggered many false positive code alerts.

#### Manual Tests

### Test Coverage






Most of this is renaming or String to Path/URI replacements. Domain objects are already covered; `EntityFileService` is covered by these.

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| Attachment.java | 94.1% |                            |
| Course.java | 93.9% |               |
| FileUploadSubmission.java | 100% |               |
| ExamUser.java | 88.7% |               |
| Slide.java | 82.1% |               |
| DragAndDropQuestion.java | 88.9% |               |
| DragItem.java | 93.5% |               |
| EntityFileService.java | 100% |               |
| FilePathService.java | 93.3% | ✅              |
| FileService.java | 85.8% | ✅              |
| QuizExerciseImportService.java | 87.9% |               |
| ProgrammingExerciseRepositoryService.java | 91.3% |             |
